### PR TITLE
Deployments

### PR DESCRIPTION
**Pull request by craig8**
# Description

This is a pull request that helps to mitigate the pain of installing volttron to multiple machines.  In general this pull request exposes the following new commands:

- vctl deploy up
- vctl deploy down
- vctl deploy status
- vctl deploy destroy

Each of the above require an ansible inventory file.  

In addition, the vctl install command now can handle a path to an agent directory as an argument.  This feature deprecates the scripts/install-agent.py script

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Tested with a 3 machine distribution created via vagrant with the specification in the examples/deployment directory

# Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by laroque** [comment]
@craig8 the last commit indicates a WIP. Are you still working on adding more to this, or is it all you have/plan to do?

**Comment by craig8** [comment]
@laroque I am planning on looking more into it in the morning...I got several of the pull requests in and updated but wanted to make sure this works for you and myself before really pushing it.  I think I pushed the wip because I had to get it up before leaving the lab!

**Comment by laroque** [comment]
@craig8 looks like you haven't had a chance to push anything up yet, have you had a chance to look yet? Could you summarize what you're still working on (the sudo issue, others?)

File: deployment/playbooks/base-install.yml [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+  vars:
+    # Ansible will use this path on remote hosts to run the interpreter with.
+    ansible_python_interpreter: /usr/bin/python3
+  tasks:
+    #- include_tasks: includes/common.yml
+
+    - name: Initialize git variables for downloading repository without git
+      set_fact:
+        git_branch: remote-platforms
+        git_organization: craig8
```
Comment (shwethanidd) [inline deployment/playbooks/base-install.yml]: This should point to 'volttron' or be made configurable

File: deployment/playbooks/includes/ansible_venv.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  pip:
+    virtualenv: "{{ my_ansible_vitualenv }}"
+    virtualenv_command: /usr/bin/python3 -m venv
+    name:
+     - pyyaml
+     - psutil
+     - zmq
+     - 'netifaces==0.10.9'
+     - 'deepdiff'
+
```
Comment (shwethanidd) [inline deployment/playbooks/includes/ansible_venv.yml]: Why do we need to install these packages in ansible virtual environment?

File: deployment/playbooks/includes/defaults.yml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+---
+- name: Set default values
+  set_fact:
+    volttron_home: "{{ volttron_home | default('{{ ansible_env.HOME }}/.volttron') }}"
+    volttron_root: "{{ volttron_root | default('{{ ansible_env.HOME }}/volttron') }}"
+    host_config: "{{ host_config | default('{{ ansible_env.HOME }}/host.config.yml') }}"
+    host_configs_dir: "{{ host_configs_dir | default('{{ ansible_env.HOME }}/configs') }}"
+    enable_web: no
```
Comment (shwethanidd) [inline deployment/playbooks/includes/defaults.yml]: Can we follow same format as above lines. This is a low priority task that Ben can take up later

File: deployment/playbooks/remove-volttron.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        - "{{ host_config }}"
+        - "{{ my_ansible_vitualenv }}"
+        - "{{ ansible_env.HOME }}/rabbitmq_server"
+        - "{{ ansible_env.HOME }}/.volttron_rmq_home"
+        - "{{ ansible_env.HOME }}/.volttron_instances"
+        - "{{ ansible_env.HOME }}/ansible_logging.log"
+        - "{{ ansible_env.HOME }}/logfile.log"
+        - "{{ ansible_env.HOME }}/configs"
+        - "{{ ansible_env.HOME }}/volttron.tar.gz"
+
```
Comment (shwethanidd) [inline deployment/playbooks/remove-volttron.yml]: Ben suggests there is more efficient to handle this. He is taking it as action item for later.

File: examples/deployment/collector2/collector2.yml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+#rabbitmq-config: $CONFIG/rabbitmq_config.yml
+
+# Agents dictionary to install.  The key must be a valid
+# identity for the agent to be installed correctly.
+agents:
+
+  # Each agent identity.config file should be in the configs
+  # directory and will be used to install the agent.
+  listener:
+    source: $VOLTTRON_ROOT/examples/ListenerAgent
```
Comment (shwethanidd) [inline examples/deployment/collector2/collector2.yml]: Refer to VOLTTRON_ROOT from volttron_environment.yml. Low priority comment.

File: deployment/playbooks/includes/ansible_venv.yml (hunk 2 above)
Comment (craig8) [reply deployment/playbooks/includes/ansible_venv.yml]: They are required...

File: deployment/playbooks/base-install.yml (hunk 1 above)
Comment (craig8) [reply deployment/playbooks/base-install.yml]: yup

File: deployment/playbooks/base-install.yml [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        # This module doesn't use the http_proxy as the env instead it uses the
+        # capitalized one
+        # reference https://docs.ansible.com/ansible/latest/modules/pip_module.html
+        HTTP_PROXY: "{{ http_proxy }}"
+      when: extra_requirments is defined and extra_requirements != []
+
+    - set_fact:
+        local_remote_config_file: "{{lookup('env','DEPLOYMENT_ROOT')}}/{{inventory_hostname}}/{{inventory_hostname}}.yml"
+        local_configs_dir: "{{lookup('env','DEPLOYMENT_ROOT')}}/{{inventory_hostname}}/configs"
+        python3: "{{lookup('env','DEPLOYMENT_ROOT')}}/env/bin/python"
```
Comment (laroque) [inline deployment/playbooks/base-install.yml]: Don't use DEPLOYMENT_ROOT root from env, this should be set in the inventory

Comment (craig8) [reply deployment/playbooks/base-install.yml]: DEPLOYMENT_ROOT is from the host perspective not the inventory perspective.  This should be the root of files to copy over to the system...note I am fine with having a different method

Comment (craig8) [reply deployment/playbooks/base-install.yml]: Note also DEPLOYMENT_ROOT is set within the vctl deploy command as the directory where the host environment is specified.  If using outside that context this should be set at least until something else changes the way this works.

Comment (laroque) [reply deployment/playbooks/base-install.yml]: I'll fix this, we're moving this to a variable which can be defined on the entire hosts group (not per host, since the logic knows to expect a subfolder per host).

**Comment by laroque** [comment]
I will:
- [ ] add a step to `init` to test if we're on a supported system and provide useful exit info if not
- [ ] write up documentation for using the various "deploy" features

**Comment by craig8** [comment]
Closing now that I think I have everything out of this put in a different location/utilization.
